internal/util/pool.go

```diff
@@ -6,105 +6,175 @@ import (
+// Always call IsDone() before accessing any results to avoid concurrency issues.
+type PoolResult[InputType comparable, OutputType any] struct {
+	// The results of the parallel pool.
+	Results map[InputType]OutputType
+
+	// A map of work errors using the item for item-level errors.
+	WorkErrors map[InputType]error
+
+	// Signals that the parallel pool was canceled during execution.
+	Canceled bool
+
+	// An internal done channel to signal the results can be accessed.
+	done chan any
+}
+
+// Use this method to block until the results can be accessed.
+func (this PoolResult[InputType, OutputType]) IsDone() {
+	<-this.done
+}
+
 // Do a map function (one result for one input) with a parallel pool of workers.
 // Unless there is a critical error (the final return value) or cancellation,
-// the output will have the same length as and index-match the input, but will have an empty value if there is an error.
-// A cancellation will return (nil, nil, nil).
-// Consult the returned map of errors using the item's index to check for item-level errors.
+// every input will either be in PoolResult.Results or PoolResult.WorkErrors.
+// A cancellation will stop new work from starting but complete in progress work asynchronously.
+// The partial results from a cancellation are available after PoolResult.IsDone() returns.
+// Consult the returned map of errors using the item to check for item-level errors.
+// Work level errors will not generate a critical error, so both must be checked independently.
 // The underlying collection of input work items must not be modified as this function is running.
-func RunParallelPoolMap[InputType any, OutputType any](poolSize int, workItems []InputType, ctx context.Context, workFunc func(InputType) (OutputType, error)) ([]OutputType, map[int]error, error) {
+func RunParallelPoolMap[InputType comparable, OutputType any](poolSize int, workItems []InputType, ctx context.Context, workFunc func(InputType) (OutputType, error)) (*PoolResult[InputType, OutputType], error) {
 	if poolSize <= 0 {
-		return nil, nil, fmt.Errorf("Pool size must be positive, got %d.", poolSize)
-	}
-
-	type WorkItem struct {
-		Index int
-		Item  InputType
+		return nil, fmt.Errorf("Pool size must be positive, got %d.", poolSize)
 	}
 
 	type ResultItem struct {
-		Index int
-		Item  OutputType
-		Error error
+		Input  InputType
+		Result OutputType
+		Error  error
 	}
 
-	results := make([]OutputType, len(workItems))
-	workErrors := make(map[int]error)
-
 	resultQueue := make(chan ResultItem, poolSize)
-	workQueue := make(chan WorkItem, poolSize)
-	doneChan := make(chan bool, poolSize)
-	exitWaitGroup := sync.WaitGroup{}
+	workQueue := make(chan InputType, poolSize)
+
+	// The done working channel is used by the pool workers to signal work is complete to the result collectors.
+	doneWorkingChan := make(chan any)
+	// The done channel is used to signal that the all results are collected and accessible.
```

Comment:
I would put a newline here to separate the two.
Although it can happen, it is less common to have comment within a visual block.

(Note that this is purely a nit, since there are other changes.)